File: FunctionalPatterns/Creational/FactoryMethod.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Creational

module FactoryMethod =

    // record is not necessary, type alias for unit -> unit would also suffice
    type FileReader = { read: unit -> unit }

    let readTextFile () = printfn "Reading text file..."

    let readPdfFile () = printfn "Reading PDF file..."

    let createFileReader (extension: string) =
        match extension.ToLower() with
        | ".txt" -> { read = readTextFile }
        | ".pdf" -> { read = readPdfFile }
        | _ -> raise (System.ArgumentException("Unsupported file type"))

File: ObjectOrientedPatterns/Creational/FactoryMethod.cs
namespace ObjectOrientedPatterns.Creational.FactoryMethod
{
    // see https://en.wikipedia.org/wiki/Factory_method_pattern

    public interface IFileReader
    {
        void Read();
    }

    class TextFileReader : IFileReader
    {
        public void Read() => Console.WriteLine("Reading text file...");
    }

    class PDFFileReader : IFileReader
    {
        public void Read() => Console.WriteLine("Reading PDF file...");
    }

    public class FileReaderFactory
    {
        public static IFileReader CreateFileReader(string fileExtension)
        {
            switch (fileExtension.ToLower())
            {
                case ".txt":
                    return new TextFileReader();
                case ".pdf":
                    return new PDFFileReader();
                default:
                    throw new ArgumentException("Unsupported file type");
            }
        }
    }
}
